pills/17/fix-function-sum.txt

```diff
@@ -0,0 +1,14 @@
+# Our old friend `sum`
+nix-repl> sum = a: b: a + b
+
+# Our new friend `fix`
+nix-repl> fix = f: let f result; in result
+
+# Helper function that forces the "fixed" function to recurse over its arguments __one argument at a time__ in each recursion
+nix-repl> const = a: b: a
+
+# Now use `fix` and `const` to turn `sum` into functions that evaluate one argument at a time:
```

Comment:
Nix only supports functions that take a single argument so the function is already [curried](https://en.wikipedia.org/wiki/Currying):

```nix
nix-repl> sum_0 = sum
nix-repl> sum_1 = sum_0 1
nix-repl> sum_1 2         
3
```

It works here because `fix (const x)` will evaluate to `x` for any `x`:

```nix
fix (const x) # expanding the definition of fix
(f: let result = f result; in result) (const x) # application of argument
let result = (const x) result; in result # expanding the definition of const
let result = ((a: b: a) x) result; in result # application of argument
let result = (b: x) result; in result # application of argument
let result = x; in result # let definition expansion
x # let definition expansion
```

Reply:
Absolutely. The problem with my example is that currying makes the example trivial rather than informative. Will fix.